docs/building-modules/msg-services.md

```diff
@@ -0,0 +1,99 @@
+and [\#7122](https://github.com/cosmos/cosmos-sdk/issues/7421).
+
+`NewHandler` function dispatches a `Msg` to appropriate `MsgServer` RPC function, usually by using a switch statement:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc1/x/bank/handler.go#L10-L30
+
+First, `NewHandler` function sets a new `EventManager` to the context to isolate events per `msg`.
+Then, a simple switch calls the appropriate `msgServer` method based on the `Msg` type. `msgServer` methods can retrieve the `sdk.Context` from the `context.Context` parameter method using the `sdk.UnwrapSDKContext`:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc1/x/bank/keeper/msg_server.go#L27
+
+`Msg` processing usually follow this 2 steps:
+
+- First, they perform *stateful* checks to make sure the `message` is valid. At this stage, the `message`'s `ValidateBasic()` method has already been called, meaning *stateless* checks on the message (like making sure parameters are correctly formatted) have already been performed. Checks performed in the `msgServer` method can be more expensive and require access to the state. For example, a `msgServer` method for a `transfer` message might check that the sending account has enough funds to actually perform the transfer. To access the state, the `msgServer` method needs to call the [`keeper`'s](./keeper.md) getter functions. 
+- Then, if the checks are successfull, the `msgServer` method calls the [`keeper`'s](./keeper.md) setter functions to actually perform the state transition. 
+
+Before returning, `msgServer` methods generally emit one or multiple [`events`](../core/events.md) via the `EventManager` held in the `ctx`:
+
+```go
+ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			eventType,  // e.g. sdk.EventTypeMessage for a message, types.CustomEventType for a custom event defined in the module
+			sdk.NewAttribute(attributeKey, attributeValue),
+		),
+    )
+```
+
+These `events` are relayed back to the underlying consensus engine and can be used by service providers to implement services around the application. Click [here](../core/events.md) to learn more about `events`. 
+
+The invoked `msgServer` method returns a `proto.Message` response and an `error`. These return values are then wrapped into an `*sdk.Result` or an `error` using `sdk.WrapServiceResult(ctx sdk.Context, res proto.Message, err error)`. This method takes care of marshaling the `res` parameter to protobuf and attaching any events on the `ctx.EventManager()` to the `sdk.Result`.
+
++++ https://github.com/cosmos/cosmos-sdk/blob/d55c1a26657a0af937fa2273b38dcfa1bb3cff9f/proto/cosmos/base/abci/v1beta1/abci.proto#L81-L95
+
+`NewHandler` can then be registered from [`AppModule.Route()`](./module-manager.md#appmodule) as shown in the example below:
```

Comment:
this shouldn't be here I think, the `Route()` is for legacy amino msgs. It's correct in the subsection below.